### PR TITLE
Deb cuda

### PR DESCRIPTION
**Pull request by tzafrir-mellanox**
This adds support for a separate ucx-cuda deb package (currently: hardwired to version 11.6).

Along the way it adds a number of required fixes to the deb packaging.

### CODE REVIEW

**Review by avildema — DISMISSED** [review]

**Review by avildema — APPROVED** [review]

File: Makefile.am [hunk 1]
```diff
@@ -62,14 +62,18 @@ endif
...
+EXTRA_DIST += contrib/configure-release-mt
 EXTRA_DIST += contrib/configure-prof
 EXTRA_DIST += contrib/buildrpm.sh
 EXTRA_DIST += contrib/ucx_perftest_config/msg_pow2
 EXTRA_DIST += contrib/ucx_perftest_config/README
 EXTRA_DIST += contrib/ucx_perftest_config/test_types_uct
 EXTRA_DIST += contrib/ucx_perftest_config/test_types_ucp
 EXTRA_DIST += contrib/ucx_perftest_config/transports
-EXTRA_DIST += debian
+EXTRA_DIST += debian/changelog debian/compat debian/copyright debian/control \
```
Comment (yosefe) [inline Makefile.am:73]: pls list each file in separate line, using `\`

File: Makefile.am [hunk 2]
```diff
@@ -62,14 +62,18 @@ endif
...
+	debian/rules debian/source/format \
+	debian/ucx.install debian/ucx-cuda.install \
+	#
```
Comment (yosefe) [inline Makefile.am:76]: pls remove this

File: debian/copyright [hunk 3]
```diff
@@ -3,13 +3,14 @@ Upstream-Name: UCX
...
-Copyright (c) 2014-2015      NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+Copyright:
+ 2014-2015      UT-Battelle, LLC. All rights reserved.
+ 2014-2015      NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
 License: BSD
  Redistribution and use in source and binary forms, with or without 
  modification, are permitted provided that the following conditions 
  are met:
- 
+ .
```
Comment (yosefe) [inline debian/copyright:13]: why is this needed?

File: debian/control.in [hunk 4]
```diff
@@ -12,9 +14,19 @@ Homepage: http://www.openucx.org
...
+
+Package: ucx-cuda
+Section: libs
+Depends: ${misc:Depends}, ${shlibs:Depends}
+Architecture: any
+Build-Profiles: <cuda>
+Description: Unified Communication X
+ UCX is a communication library implementing high-performance messaging.
+ .
+ This package includes CUDA support for UCX.
```
Comment (yosefe) [inline debian/control.in:32]: Provide CUDA (NVIDIA GPU) support for UCX. Enables passing GPU memory pointers
to UCX communication routines, and transports taking advantage of GPU-Direct
technology for direct data transfer between GPU and RDMA devices.

File: debian/rules.in [hunk 5]
```diff
@@ -9,14 +9,36 @@
 # Uncomment this to turn on verbose mode.
 #export DH_VERBOSE=1
 
+CUDA_VER = 11.6
```
Comment (yosefe) [inline debian/rules.in:12]: why need a specific version?

File: debian/rules.in [hunk 6]
```diff
@@ -9,14 +9,36 @@
...
+saved_files = changelog control rules
 override_dh_auto_clean:
+	for f in $(saved_files); do \
+	  cp -a debian/$$f debian/$$f.sav; \
+	done
+	dh_auto_clean
+	for f in $(saved_files); do \
+	  if [ ! -r debian/$$f ]; then mv -v debian/$$f.sav debian/$$f; fi; \
+	  rm -f debian/$$f.sav; \
+	done
```
Comment (yosefe) [inline debian/rules.in:44]: why needed?

File: debian/ucx.install [hunk 7]
```diff
@@ -0,0 +1,9 @@
+usr/bin/*
+usr/include/*
+usr/lib/lib*
+usr/lib/cmake
+usr/lib/pkgconfig
+usr/lib/ucx/libuct_cma.*
```
Comment (yosefe) [inline debian/ucx.install:6]: also %{_libdir}/ucx/libuct_rdmacm.so.*

File: debian/copyright (hunk 3 above)
Comment (tzafrir-mellanox) [reply debian/copyright:13]: This is part of the format. It is also used in the description in debian/control .
See https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#formatted-text
https://www.debian.org/doc/debian-policy/ch-controlfields#description

File: debian/rules.in (hunk 6 above)
Comment (tzafrir-mellanox) [reply debian/rules.in:44]: That was a hack because of files getting deleted by the clean target.

Comment (yosefe) [reply debian/rules.in:44]: maybe update the clean target? or prevent it from being run?
is this a standard method for deb?

Comment (tzafrir-mellanox) [reply debian/rules.in:44]: The standard in Debian is that the clean target is run. And does not delete any essential deb packaging files.

Comment (yosefe) [reply debian/rules.in:44]: any idea why this happens? i couldn't find any debian clean steps in Makefile.am..

File: debian/rules.in (hunk 5 above)
Comment (tzafrir-mellanox) [reply debian/rules.in:12]: Got rid of that. But there's still the build dependency in debian/control .

**Comment by tzafrir-mellanox** [comment]
Running
    debian/rules clean

runs:  dh clean

    $ dh clean --no-act
       dh_testdir
       debian/rules override_dh_auto_clean
       dh_clean

https://manpages.debian.org/bullseye/debhelper/dh_auto_clean.1.en.html
So it runs 'make distclean' in this case.

File: debian/rules.in (hunk 5 above)
Comment (yosefe) [reply debian/rules.in:12]: could we require a lower version? such as 11.1?

File: Makefile.am (hunk 1 above)
Comment (tzafrir-mellanox) [reply Makefile.am]: Fixed

File: debian/rules.in [hunk 8]
```diff
@@ -9,14 +9,34 @@
...
+# We may have a single package or two packages. We can't tell in advance.
+# Therefore set destdir for the default of multiple packages (that also
+# works for a single package):
+override_dh_auto_install:
+	dh_auto_install --destdir=debian/tmp
+
 override_dh_shlibdeps:
 	dh_shlibdeps --dpkg-shlibdeps-params=--ignore-missing-info
 
+saved_files = changelog control rules
```
Comment (yosefe) [inline debian/rules.in:33]: can you please add a comment like
`# dh_auto_clean runs 'make distclean' that removes files generated by configure script, but we still need them`

File: debian/control.in [hunk 9]
```diff
@@ -1,9 +1,13 @@
...
 Section: libs
 Priority: extra
 Maintainer: ucx-group@elist.ornl.gov
-Build-Depends:
-Standards-Version: @MAJOR_VERSION@.@MINOR_VERSION@
+Build-Depends: debhelper (>= 9.0),
+ libibverbs-dev,
+ librdmacm-dev,
+ pkg-config,
+ cuda-libraries-dev-11-6 <cuda>,
```
Comment (yosefe) [inline debian/control.in:9]: it means it won't build with any other version, even newer?

File: debian/control.in [hunk 10]
```diff
@@ -1,9 +1,13 @@
...
 Section: libs
 Priority: extra
 Maintainer: ucx-group@elist.ornl.gov
-Build-Depends:
-Standards-Version: @MAJOR_VERSION@.@MINOR_VERSION@
+Build-Depends: debhelper (>= 9.0),
+ libibverbs-dev,
+ librdmacm-dev,
+ pkg-config,
+# cuda-libraries-dev-11-6 <cuda>,
```
Comment (yosefe) [inline debian/control.in:9]: remove?

**Review by yosefe — DISMISSED** [review]
pls squash

**Review by yosefe — APPROVED** [review]
